core/domain/draft_upgrade_services.py

```diff
@@ -0,0 +1,82 @@
+# coding: utf-8
+#
+# Copyright 2019 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Commands that can be used to upgrade draft to newer Exploration versions."""
+import logging
+
+from core.domain import exp_domain
+from core.platform import models
+
+(exp_models, feedback_models, user_models) = models.Registry.import_models([
+    models.NAMES.exploration, models.NAMES.feedback, models.NAMES.user
+])
+
+
+def try_upgrade_from_version_to_version(
+        draft_change_list, from_version, to_version, exp_id):
+    """Try upgrading an ExplorationChange domain object by calling convert
+    functions for a list of commits.
+
+    For now, this handles only the scenario where commits_list contains a
+    single commit that migrates the state schema.
+
+    Args:
+        draft_change_list: list(dict). Each dict can be parsed as an
```

Comment:
Can you pass domain objects instead? That would be better.

Reply:
If we pass in domain objects, we would have to read draft_change_list field, parse into domain objects, call draft upgrade, and then call toDict() inorder to save the new draft_change_list. Is there any reason we prefer domain objects?

Reply:
Yes, that is the intended flow. The problem with passing dicts around is that their fields can be mutated too easily -- they lack structure. E.g. you could add a key-value pair midway and no one would be any the wiser. That's why we use domain objects with named fields throughout the codebase.

Conversion to dict should happen only at the application boundaries -- i.e. when the app talks to the frontend, or when we need to convert to dict for storage purposes. But internally we should be using domain objects throughout. They are essentially DTOs ("data transfer objects").

Reply:
Ahh that makes sense. Done.